### PR TITLE
added option to trimm the trace to a monotonically decreasing one

### PR DESCRIPTION
**Pull request by PaulJonasJost**
I wrote a function to get only a monotonically decreasing history trace. This is meant for comparisons with e.g. test and training data but can obviously also be used for more.

It does not change any functionality from before.

### CODE REVIEW

File: pypesto/objective/history.py [hunk 1]
```diff
@@ -24,13 +24,16 @@ def trace_wrap(f):
     index list, and reduces for integer `ix` the output to a single value.
     """
     def wrapped_f(
-            self, ix: Union[Sequence[int], int, None] = None
+            self, ix: Union[Sequence[int], int, None] = None,
+            trimm: bool = False
```
Comment (dilpath) [inline pypesto/objective/history.py:28]:
```suggestion
            trim: bool = False
```

File: pypesto/objective/history.py [hunk 2]
```diff
@@ -319,6 +331,20 @@ def get_time_trace(
...
+        decreasing history.
+        """
+        fval_trace = self.get_fval_trace()
+        indices = []
+        fval_current = np.inf
+        for iter_i, fval_i in enumerate(fval_trace):
+            if fval_i <= fval_current:
+                indices.append(iter_i)
+                fval_current = fval_i
+        return indices
```
Comment (dweindl) [inline pypesto/objective/history.py:346]:
```suggestion
        return np.where(fval_trace <= np.minimum.accumulate(fval_trace))
```

**Review by dweindl — COMMENTED** [review]
Add a test and use in some notebook?

File: pypesto/objective/history.py (hunk 2 above)
Comment (PaulJonasJost) [reply pypesto/objective/history.py:346]: wow that looks really elegant :)

File: pypesto/objective/history.py [hunk 3]
```diff
@@ -24,13 +24,16 @@ def trace_wrap(f):
...
+            self, ix: Union[Sequence[int], int, None] = None,
+            trim: bool = False
     ) -> Union[Sequence[Union[float, MaybeArray]], Union[float, MaybeArray]]:
         # whether to reduce the output
         reduce = isinstance(ix, numbers.Integral)
         # default: full list
         if ix is None:
             ix = np.arange(0, len(self), dtype=int)
+            if trim:
+                ix = self.get_trimmed_indices()
```
Comment (dweindl) [inline pypesto/objective/history.py:36]:
```suggestion
            if trim:
                ix = self.get_trimmed_indices()
            else:
                ix = np.arange(0, len(self), dtype=int)
```

File: test/base/test_history.py [hunk 4]
```diff
@@ -620,3 +620,30 @@ def test_hdf5_history_mp():
...
+    """
+    problem = CRProblem()
+    pypesto_problem = problem.get_problem()
+
+    optimizer = pypesto.optimize.ScipyOptimizer()
+    history_options = pypesto.HistoryOptions(trace_record=True)
+    result = pypesto.optimize.minimize(
+        problem=pypesto_problem, optimizer=optimizer,
+        n_starts=2, history_options=history_options,
+        engine=MultiProcessEngine()
```
Comment (dweindl) [inline test/base/test_history.py:638]: Why launch two optimizations if only one is used later on?

**Review by dweindl — APPROVED** [review]

File: test/base/test_history.py (hunk 4 above)
Comment (PaulJonasJost) [reply test/base/test_history.py:638]: you have a point there, I had a reason but I cannot remember what it was :D

File: test/base/test_history.py [hunk 5]
```diff
@@ -620,3 +620,30 @@ def test_hdf5_history_mp():
...
+    decreasing.
+    """
+    problem = CRProblem()
+    pypesto_problem = problem.get_problem()
+
+    optimizer = pypesto.optimize.ScipyOptimizer()
+    history_options = pypesto.HistoryOptions(trace_record=True)
+    result = pypesto.optimize.minimize(
+        problem=pypesto_problem, optimizer=optimizer,
+        n_starts=2, history_options=history_options,
```
Comment (PaulJonasJost) [inline test/base/test_history.py:637]:
```suggestion
        n_starts=1, history_options=history_options,
```

**Pull request merged**